File: packages/engine/src/book.ts
import BTreeImport from 'sorted-btree';
import type { Order, Side } from '@exchange/common';

// sorted-btree is CJS; under Node's native ESM loader the default import
// sometimes resolves to the module namespace object rather than the class.
// vitest's transform hides this; Node does not. Normalize once here.
const BTree = (
  (BTreeImport as unknown as { default?: typeof BTreeImport }).default ??
  BTreeImport
) as typeof BTreeImport;

/**
 * Order book data structure.
 *
 * Two sorted maps (B-trees) keyed by price (bigint):
 *   - bids: highest price first (best bid = first)
 *   - asks: lowest price first  (best ask = first)
 *
 * Each price level holds a FIFO doubly-linked list of resting orders.
 * Time priority is preserved by always appending new orders at the TAIL
 * and matching from the HEAD.
 *
 * We also keep an `orderId -> { node, level, side }` map so cancels are
 * O(log n) (price lookup + O(1) unlink) instead of scanning.
 *
 * Why a DLL inside each level? It gives us:
 *   - O(1) head pop during matching
 *   - O(1) unlink on cancel (we have the node ref via the map)
 *   - Time order without resorting on every insert
 */

interface ListNode {
  order: Order;
  prev: ListNode | null;
  next: ListNode | null;
}

export class PriceLevel {
  head: ListNode | null = null;
  tail: ListNode | null = null;
  totalQty: bigint = 0n;
  count: number = 0;

  append(order: Order): ListNode {
    const node: ListNode = { order, prev: this.tail, next: null };
    if (this.tail) this.tail.next = node;
    else this.head = node;
    this.tail = node;
    this.totalQty += order.remaining;
    this.count += 1;
    return node;
  }

  unlink(node: ListNode): void {
    if (node.prev) node.prev.next = node.next;
    else this.head = node.next;
    if (node.next) node.next.prev = node.prev;
    else this.tail = node.prev;
    this.totalQty -= node.order.remaining;
    this.count -= 1;
  }

  isEmpty(): boolean {
    return this.head === null;
  }
}

interface OrderRef {
  node: ListNode;
  level: PriceLevel;
  side: Side;
  price: bigint;
}

export class OrderBook {
  // Bids: highest price first. Comparator returns negative when a > b.
  readonly bids = new BTree<bigint, PriceLevel>(
    undefined,
    (a, b) => (a > b ? -1 : a < b ? 1 : 0)
  );
  // Asks: lowest price first. Default ascending comparator on bigint.
  readonly asks = new BTree<bigint, PriceLevel>(
    undefined,
    (a, b) => (a < b ? -1 : a > b ? 1 : 0)
  );
  private readonly refs = new Map<string, OrderRef>();

  bestBid(): bigint | undefined {
    const r = this.bids.minKey();
    return r === undefined ? undefined : r;
  }

  bestAsk(): bigint | undefined {
    const r = this.asks.minKey();
    return r === undefined ? undefined : r;
  }

  /** Add a resting order. Caller must have decided this order should rest. */
  rest(order: Order): void {
    const tree = order.side === 'buy' ? this.bids : this.asks;
    let level = tree.get(order.price);
    if (!level) {
      level = new PriceLevel();
      tree.set(order.price, level);
    }
    const node = level.append(order);
    this.refs.set(order.orderId, {
      node,
      level,
      side: order.side,
      price: order.price,
    });
  }

  /**
   * Pop the head (best, oldest) order on the given side. Returns undefined
   * if that side is empty. Used by the matcher.
   */
  peekTopOrder(side: Side): Order | undefined {
    const tree = side === 'buy' ? this.bids : this.asks;
    const topPrice = tree.minKey();
    if (topPrice === undefined) return undefined;
    const level = tree.get(topPrice)!;
    return level.head?.order;
  }

  /**
   * Reduce or remove the head order on `side`. If qty == head.remaining,
   * the order is unlinked and removed from refs. Otherwise its `remaining`
   * is decreased and `totalQty` on the level is decreased.
   */
  reduceTopOrder(side: Side, fillQty: bigint): void {
    const tree = side === 'buy' ? this.bids : this.asks;
    const topPrice = tree.minKey();
    if (topPrice === undefined) throw new Error('reduceTopOrder on empty side');
    const level = tree.get(topPrice)!;
    const node = level.head;
    if (!node) throw new Error('reduceTopOrder on empty level');
    if (fillQty > node.order.remaining) {
      throw new Error('reduceTopOrder fillQty > remaining');
    }
    if (fillQty === node.order.remaining) {
      this.refs.delete(node.order.orderId);
      level.unlink(node);
      if (level.isEmpty()) tree.delete(topPrice);
    } else {
      node.order.remaining -= fillQty;
      level.totalQty -= fillQty;
    }
  }

  /** Cancel a resting order by id. Returns the canceled order or undefined. */
  cancel(orderId: string): Order | undefined {
    const ref = this.refs.get(orderId);
    if (!ref) return undefined;
    this.refs.delete(orderId);
    ref.level.unlink(ref.node);
    const tree = ref.side === 'buy' ? this.bids : this.asks;
    if (ref.level.isEmpty()) tree.delete(ref.price);
    return ref.node.order;
  }

  /** Look up a resting order (does not mutate). */
  get(orderId: string): Order | undefined {
    return this.refs.get(orderId)?.node.order;
  }

  /** Snapshot of top-N depth for a side: [price, totalQty][]. */
  depth(side: Side, levels: number): [bigint, bigint][] {
    const tree = side === 'buy' ? this.bids : this.asks;
    const out: [bigint, bigint][] = [];
    let count = 0;
    for (const [price, level] of tree.entries()) {
      out.push([price, level.totalQty]);
      if (++count >= levels) break;
    }
    return out;
  }
}

File: packages-go/engine/book.go
package main

import (
	"errors"

	"github.com/KaustubhPatange/exchange/common"
	"github.com/google/btree"
)

type listNode struct {
	Order *common.Order
	Prev  *listNode
	Next  *listNode
}

type priceLevel struct {
	Head     *listNode
	Tail     *listNode
	TotalQty int64
	Count    int
}

func (p *priceLevel) Append(order *common.Order) *listNode {
	node := &listNode{Order: order, Prev: p.Tail, Next: nil}
	if p.Tail != nil {
		p.Tail.Next = node
	} else {
		p.Head = node
	}
	p.Tail = node
	p.TotalQty += order.Remaining
	p.Count += 1
	return node
}

func (p *priceLevel) Unlink(node *listNode) {
	if node.Prev != nil {
		node.Prev.Next = node.Next
	} else {
		p.Head = node.Next
	}
	if node.Next != nil {
		node.Next.Prev = node.Prev
	} else {
		p.Tail = node.Prev
	}
	p.TotalQty -= node.Order.Remaining
	p.Count -= 1
}

func (p *priceLevel) IsEmpty() bool {
	return p.Head == nil
}

type OrderRef struct {
	Node  *listNode
	Level *priceLevel
	Side  common.Side
	Price int64
}

type priceEntry struct {
	price int64
	level *priceLevel
}

type OrderBook struct {
	Bids *btree.BTreeG[priceEntry]
	Asks *btree.BTreeG[priceEntry]
	refs map[string]*OrderRef
}

func NewOrderBook() *OrderBook {
	return &OrderBook{
		Bids: btree.NewG(32, func(a, b priceEntry) bool {
			return a.price > b.price
		}),
		Asks: btree.NewG(32, func(a, b priceEntry) bool {
			return a.price < b.price
		}),
		refs: make(map[string]*OrderRef),
	}
}

func (o *OrderBook) treeFor(side common.Side) *btree.BTreeG[priceEntry] {
	if side == common.SideBuy {
		return o.Bids
	} else {
		return o.Asks
	}
}

func (o *OrderBook) BestBid() (int64, bool) {
	r, ok := o.Bids.Min()
	if !ok {
		return 0, false
	}
	return r.price, true
}

func (o *OrderBook) BestAsk() (int64, bool) {
	r, ok := o.Asks.Min()
	if !ok {
		return 0, false
	}
	return r.price, true
}

func (o *OrderBook) Rest(order *common.Order) {
	tree := o.treeFor(order.Side)
	entry, ok := tree.Get(priceEntry{price: order.Price})
	var level *priceLevel
	if !ok {
		level = &priceLevel{}
		tree.ReplaceOrInsert(priceEntry{price: order.Price, level: level})
	} else {
		level = entry.level
	}
	node := level.Append(order)
	o.refs[order.OrderID] = &OrderRef{
		Node:  node,
		Level: level,
		Side:  order.Side,
		Price: order.Price,
	}
}

func (o *OrderBook) PeekTopOrder(side common.Side) *common.Order {
	tree := o.treeFor(side)
	top, ok := tree.Min()
	if !ok {
		return nil
	}
	if top.level.Head != nil {
		return nil
	}
	return top.level.Head.Order
}

func (o *OrderBook) ReduceTopOrder(side common.Side, fillQty int64) error {
	tree := o.treeFor(side)
	top, ok := tree.Min()
	if !ok {
		return errors.New("reduceTopOrder on empty side")
	}

	level := top.level
	node := level.Head
	if node == nil {
		return errors.New("reduceTopOrder on empty level")
	}

	if fillQty > node.Order.Remaining {
		return errors.New("reduceTopOrder fillQty > remaining")
	}

	if fillQty == node.Order.Remaining {
		delete(o.refs, node.Order.OrderID)
		level.Unlink(node)
		if level.IsEmpty() {
			tree.Delete(top)
		}
	} else {
		node.Order.Remaining -= fillQty
		level.TotalQty -= fillQty
	}

	return nil
}

func (o *OrderBook) Cancel(orderID string) *common.Order {
	ref, ok := o.refs[orderID]
	if !ok {
		return nil
	}
	delete(o.refs, orderID)
	ref.Level.Unlink(ref.Node)

	tree := o.treeFor(ref.Side)
	if ref.Level.IsEmpty() {
		tree.Delete(priceEntry{price: ref.Price})
	}
	return ref.Node.Order
}

func (o *OrderBook) Get(orderID string) *common.Order {
	ref, ok := o.refs[orderID]
	if !ok {
		return nil
	}
	return ref.Node.Order
}

func (o *OrderBook) OrdersForUser(userID string) []*common.Order {
	var out []*common.Order
	for _, tree := range []*btree.BTreeG[priceEntry]{o.Bids, o.Asks} {
		tree.Ascend(func(p priceEntry) bool {
			for node := p.level.Head; node != nil; node = node.Next {
				if node.Order.UserID == userID {
					out = append(out, node.Order)
				}
			}
			return true
		})
	}
	return out
}

type DepthLevel struct {
	Price    int64
	TotalQty int64
}

func (o *OrderBook) Depth(side common.Side, levels int) []DepthLevel {
	tree := o.treeFor(side)
	out := make([]DepthLevel, 0, levels)
	count := 0
	tree.Ascend(func(item priceEntry) bool {
		out = append(out, DepthLevel{Price: item.price, TotalQty: item.level.TotalQty})
		count++
		return count < levels
	})
	return out
}
